### PR TITLE
8329150: Remove CDS support for LatestMethodCache

### PR DESCRIPTION
**Pull request by iklam**
The CDS support for LatestMethodCache has no meaningful performance benefit. I am removing it for simplcity.

I also cleaned up the code a bit
- No need to put LatestMethodCache in universe.hpp.
- No need to use TRAPS in the init function
- No need to malloc the LatestMethodCache objects

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329150](https://bugs.openjdk.org/browse/JDK-8329150): Remove CDS support for LatestMethodCache (**Enhancement** - P4)

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - Committer) ⚠️ Review applies to [422b4ec1](https://git.openjdk.org/jdk/pull/18516/files/422b4ec1b49ff239a3d1e5fa593aabd807e7824b)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18516/head:pull/18516` \
`$ git checkout pull/18516`

Update a local copy of the PR: \
`$ git checkout pull/18516` \
`$ git pull https://git.openjdk.org/jdk.git pull/18516/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18516`

View PR using the GUI difftool: \
`$ git pr show -t 18516`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18516.diff">https://git.openjdk.org/jdk/pull/18516.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18516#issuecomment-2023077554)

### CODE REVIEW

File: src/hotspot/share/memory/universe.hpp [hunk 1]
```diff
@@ -115,11 +89,11 @@ class Universe: AllStatic {
...
-  static LatestMethodCache* _finalizer_register_cache; // static method for registering finalizable objects
-  static LatestMethodCache* _loader_addClass_cache;    // method for registering loaded classes in class loader vector
-  static LatestMethodCache* _throw_illegal_access_error_cache; // Unsafe.throwIllegalAccessError() method
-  static LatestMethodCache* _throw_no_such_method_error_cache; // Unsafe.throwNoSuchMethodError() method
-  static LatestMethodCache* _do_stack_walk_cache;      // method for stack walker callback
+  static LatestMethodCache _finalizer_register_cache; // static method for registering finalizable objects
+  static LatestMethodCache _loader_addClass_cache;    // method for registering loaded classes in class loader vector
+  static LatestMethodCache _throw_illegal_access_error_cache; // Unsafe.throwIllegalAccessError() method
+  static LatestMethodCache _throw_no_such_method_error_cache; // Unsafe.throwNoSuchMethodError() method
+  static LatestMethodCache _do_stack_walk_cache;      // method for stack walker callback
```
Comment (coleenp) [inline src/hotspot/share/memory/universe.hpp:96]: Does this compile with all compilers with the forward declaration if the type isn't a pointer?  You could further reduce this by moving these static declarations into universe.cpp.  I don't see why they have to be exposed here.

File: src/hotspot/share/memory/universe.cpp [hunk 2]
```diff
@@ -133,11 +154,11 @@ OopHandle Universe::_virtual_machine_error_instance;
...
-LatestMethodCache* Universe::_finalizer_register_cache = nullptr;
-LatestMethodCache* Universe::_loader_addClass_cache    = nullptr;
-LatestMethodCache* Universe::_throw_illegal_access_error_cache = nullptr;
-LatestMethodCache* Universe::_throw_no_such_method_error_cache = nullptr;
-LatestMethodCache* Universe::_do_stack_walk_cache     = nullptr;
+LatestMethodCache Universe::_finalizer_register_cache;
+LatestMethodCache Universe::_loader_addClass_cache;
+LatestMethodCache Universe::_throw_illegal_access_error_cache;
+LatestMethodCache Universe::_throw_no_such_method_error_cache;
+LatestMethodCache Universe::_do_stack_walk_cache;
```
Comment (coleenp) [inline src/hotspot/share/memory/universe.cpp:161]: Just make these LatestMethodCache and not static data members in Universe, and move them up with the LatestMethodCache type declaration so there're all in one place.

**Review by coleenp — CHANGES_REQUESTED** [review]
This looks a nice cleanup, but I have a couple of additional suggestions.

File: src/hotspot/share/memory/universe.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/memory/universe.cpp]: Done. Thanks for the suggestion.

File: src/hotspot/share/memory/universe.cpp [hunk 3]
```diff
@@ -1255,23 +1239,27 @@ uintptr_t Universe::verify_mark_bits() {
...
+    // <init> function before java_lang_Class is linked. Print error and exit.
+    vm_exit_during_initialization(err_msg("Unable to link/verify %s.%s method",
+                                 ik->name()->as_C_string(), method));
   }
-#endif
 
+  _klass = ik;
   _method_idnum = m->method_idnum();
   assert(_method_idnum >= 0, "sanity check");
 }
```
Comment (coleenp) [inline src/hotspot/share/memory/universe.cpp:1261]: Can you move this function up to where the accessor are too?   Like around line 930?

**Review by coleenp — APPROVED** [review]
One more suggestion.  The functions were spread across universe.cpp and they really should be together.

File: src/hotspot/share/memory/universe.cpp (hunk 3 above)
Comment (iklam) [reply src/hotspot/share/memory/universe.cpp]: Done

File: src/hotspot/share/memory/universe.cpp [hunk 4]
```diff
@@ -87,6 +87,33 @@
...
+
+  void   init(JavaThread* current, InstanceKlass* ik, const char* method,
+              Symbol* signature, bool is_static);
+  Klass* klass() const           { return _klass; }
+  int    method_idnum() const    { return _method_idnum; }
+
+  Method* get_method();
+};
+
+static LatestMethodCache _finalizer_register_cache; // static method for registering finalizable objects
```
Comment (matias9927) [inline src/hotspot/share/memory/universe.cpp:111]: Could you align these comments?

**Review by matias9927 — APPROVED** [review]
Fix looks good, thanks! I have one format comment but nothing major.

File: src/hotspot/share/memory/universe.cpp (hunk 4 above)
Comment (iklam) [reply src/hotspot/share/memory/universe.cpp]: I aligned the comments and simplified the text.

File: src/hotspot/share/memory/universe.cpp [hunk 5]
```diff
@@ -924,55 +923,69 @@ void universe_oopstorage_init() {
...
-  method_cache->init(ik, m);
+
+  _klass = ik;
+  _method_idnum = m->method_idnum();
+  assert(_method_idnum >= 0, "sanity check");
+}
+
+Method* LatestMethodCache::get_method() {
+  if (klass() == nullptr) return nullptr;
+  InstanceKlass* ik = InstanceKlass::cast(klass());
```
Comment (coleenp) [inline src/hotspot/share/memory/universe.cpp:949]: Sorry for the bits and pieces.  Hopefully the last thing.  Can _klass be a InstanceKlass so we don't need this cast?

Comment (iklam) [reply src/hotspot/share/memory/universe.cpp]: Thanks for the suggestions. This code is very old and really needs to be cleaned up. I also removed the unnecessary accessor functions.

**Review by coleenp — APPROVED** [review]
It is old code.  Thanks for the cleanup.

**Comment by iklam** [comment]
Thanks @coleenp and @matias9927 for the review
/integrate

**Pull request closed without merging**